pipeline.py

```diff
@@ -178,28 +183,48 @@
     # Note: This should happen as late as possible in order to reduce the risk of the remainder of the pipeline failing
     # after a Drive upload has occurred. Failures could result in inconsistent outputs or outputs with no
     # traced data log.
-    if drive_upload:
+    if pipeline_configuration.drive_upload is not None:
         print("Uploading CSVs to Google Drive...")
-        drive_client_wrapper.init_client(drive_credentials_path)
 
-        csv_by_message_drive_dir = os.path.dirname(csv_by_message_drive_path)
-        csv_by_message_drive_file_name = os.path.basename(csv_by_message_drive_path)
-        drive_client_wrapper.update_or_create(csv_by_message_output_path, csv_by_message_drive_dir,
-                                              target_file_name=csv_by_message_drive_file_name,
-                                              target_folder_is_shared_with_me=True)
+        # Fetch the Rapid Pro Token from the Google Cloud Storage URL
+        parsed_rapid_pro_token_file_url = urlparse(pipeline_configuration.drive_upload.drive_credentials_file_url)
+        bucket_name = parsed_rapid_pro_token_file_url.netloc
+        blob_name = parsed_rapid_pro_token_file_url.path.lstrip("/")
 
-        csv_by_individual_drive_dir = os.path.dirname(csv_by_individual_drive_path)
-        csv_by_individual_drive_file_name = os.path.basename(csv_by_individual_drive_path)
-        drive_client_wrapper.update_or_create(csv_by_individual_output_path, csv_by_individual_drive_dir,
-                                              target_file_name=csv_by_individual_drive_file_name,
-                                              target_folder_is_shared_with_me=True)
+        print(f"Downloading Drive service account credentials from file '{blob_name}' in bucket '{bucket_name}'...")
```

Comment:
I think this should be moved earlier. If the service account credentials fail, we shouldn't run the whole of the pipeline.

I agree the upload should happen last, but the prerequisites should be checked at the beginning of the pipeline.

Reply:
Moved